src/update-docker-tags.ts

```diff
@@ -0,0 +1,183 @@
+
+    // FIXME error handling
+    const equivalentTags = (
+      await dockerRegistryClient.getAllEquivalentTags({
+        dockerImageRepository: trackable.dockerImageRepository,
+        tag: trackable.trackMutableTag,
+      })
+    ).filter((t) => t.startsWith(prefix));
+
+    // We assume that all the tags with the triple-dash in them are immutable:
+    // once they point at a particular SHA, they never change. (Whereas the tag
+    // we're "tracking" is mutable.)
+    //
+    // If the current tag has the right format for an immutable tag and it
+    // points to the same image as the mutable tag, leave it alone: there's no
+    // reason to create a no-op diff.
+    if (equivalentTags.includes(trackable.tag)) {
+      core.info(
+        `for image ${trackable.dockerImageRepository}:${trackable.trackMutableTag}, preserving current tag ${trackable.tag}`,
+      );
+      continue;
+    }
+    // We can choose *any* of these equivalent triple-dashed tags, and it will
+    // select the correct image version.
+    //
+    // Our tag structure increase over time (by including the number of commits
+    // since the start as determined by `git rev-list --first-parent --count
+    // HEAD`), and also includes the git commit at which it was built.
+    //
+    // So by choosing the lexicographically earliest of the equivalent tags, we
+    // are most likely to choose a tag where the named git commit actually is
+    // the commit that made a relevant change that affected the image.
+    //
+    // Additionally, this means that reverting a code change is likely to result
+    // in a revert of the tag.  Imagine that tag `main---00123-abcd` is
+    // currently running in both staging and prod, and a code change moves
+    // `main` to `main---00130-bcde` and this is deployed to staging (opening a
+    // prod promotion PR). A bug is found, so the code change X is reverted.
+    // Reproducible builds will mean that the newest tag `main---00134-dcba`
+    // will hopefully point to the same image version as `main---00123-abcd`.
+    // Choosing the min version here will mean that we will in fact "revert"
+    // staging to `main---00123-abcd`. This is now the exact same tag that is
+    // running in prod, so the prod promotion PR can auto-close rather than
+    // encouraging us to consider a no-op deploy to prod.
+    const earliestMatchingTag = min(
+      equivalentTags.filter((t) => t.startsWith(prefix)),
+    );
+    if (!earliestMatchingTag) {
+      throw new Error(
+        `No tags on ${trackable.dockerImageRepository} start with '${prefix}'`,
+      );
+    }
+
+    // It's OK if the current one is null because that's what we're overwriting, but we shouldn't
+    // overwrite *to* something that doesn't exist.
+    core.info(
+      `for image ${trackable.dockerImageRepository}:${trackable.trackMutableTag}, selecting earliest label ${earliestMatchingTag}`,
+    );
+    if (trackable.tag === earliestMatchingTag) {
```

Comment:
nit: This is theoretically handled by the earlier `equivalentTags.includes(trackable.tag))`, but may be nice to keep it anyway?

Reply:
Nah, happy to simplify. I didn't clean this up when I added to earlier conditional.